File: apis/core-api/internal/features/saveditems/wire_contract_test.go
package saveditems

import (
	"encoding/json"
	"sort"
	"testing"

	"github.com/stretchr/testify/require"
)

// ── The cross-language wire contract (033, following 028's pattern) ─────────────────────────────
//
// SAVED_ITEM_WIRE_JSON below is duplicated BYTE-FOR-BYTE in the Kotlin half at
// apps/customer-mobile/.../features/saved/SavedWireContractTest.kt. Neither side generates it and
// neither imports it. That is the point — a shared fixture moves with the bug, so both copies are
// maintained by hand and a divergence shows up as a failure rather than as agreement.
//
// ⚠ 027 lost days to a defect no unit test could see: Kotlin serialised quantities as Double, so the
// wire carried `1.0`, and Go's encoding/json cannot unmarshal `1.0` into an int. Every test passed
// throughout, because the fakes spoke Kotlin at both ends and never crossed the wire. This test is
// what would have caught it on day one.
//
// ⚠ 029's lesson about the fixture itself: its banner_test.go ASSERTED THE DEFECT, demanding
// Kind == "search" and pinning {"kind":"sale"} — a shape no banner ever emitted. A fixture that
// agrees with the code rather than with the world proves nothing. The literal below is the marshalled
// output of a real handler DTO, verified against it by TestWireLiteralIsWhatTheHandlerActuallyEmits.

const SAVED_ITEM_WIRE_JSON = `{"id":"9f2c1d4e-0000-0000-0000-000000000001","name":"Free Range Eggs 12pk","brand":"Effy","imageUrl":"https://media.example/eggs.jpg","priceAmount":"6.50","currency":"AUD","compareAtAmount":"8.00","badges":["on_sale"],"savedAt":"2026-07-20T04:11:00Z","savedPriceAmount":"8.00","priceDropped":true,"verdict":"purchasable","categoryKey":"dairy-eggs"}`

const SAVED_MEMBERSHIP_WIRE_JSON = `{"productIds":["9f2c1d4e-0000-0000-0000-000000000001"],"count":1}`

func wireFixture() savedItemDTO {
	brand, compare, key := "Effy", "8.00", "dairy-eggs"
	img := "https://media.example/eggs.jpg"
	return savedItemDTO{
		ProductID:       "9f2c1d4e-0000-0000-0000-000000000001",
		Name:            "Free Range Eggs 12pk",
		Brand:           &brand,
		ImageURL:        &img,
		PriceAmount:     "6.50",
		Currency:        "AUD",
		CompareAtAmount: &compare,
		Badges:          []string{"on_sale"},
		SavedAt:         "2026-07-20T04:11:00Z",
		SavedPrice:      "8.00",
		PriceDropped:    true,
		Verdict:         VerdictPurchasable,
		CategoryKey:     &key,
	}
}

// TestWireLiteralIsWhatTheHandlerActuallyEmits keeps the literal honest.
//
// ⚠ This is the guard against 029's mistake. Without it the literal is just a string someone wrote,
// and it can drift into pinning a shape the handler never produces — at which point both this test
// and its Kotlin twin pass while the real payload is something else.
func TestWireLiteralIsWhatTheHandlerActuallyEmits(t *testing.T) {
	got, err := json.Marshal(wireFixture())
	require.NoError(t, err)
	require.JSONEq(t, SAVED_ITEM_WIRE_JSON, string(got),
		"the fixture must be the handler's real output — if you changed a json tag, update BOTH this "+
			"literal and the byte-identical copy in SavedWireContractTest.kt")
}

// TestSavedItemWireKeys pins the exact key set.
//
// A silent `json:"..."` rename compiles fine on both sides and breaks nothing until a client reads a
// field that is suddenly absent. Only a key-set comparison catches it.
func TestSavedItemWireKeys(t *testing.T) {
	var m map[string]json.RawMessage
	require.NoError(t, json.Unmarshal([]byte(SAVED_ITEM_WIRE_JSON), &m))

	got := make([]string, 0, len(m))
	for k := range m {
		got = append(got, k)
	}
	sort.Strings(got)

	// Transcribed from packages/shared-types/src/saved-item.ts `SavedItemDTO`.
	//
	// ⚠ THIS LIST IS COPIED FROM THE CONTRACT, NOT FROM THE STRUCT ABOVE, and that distinction is
	// load-bearing. Written from the struct it can only ever agree with itself — which is how the
	// first version of this test passed while `SavedItemDTO` still extended StorefrontProductCardDTO
	// and therefore required an `available` field the handler never emitted. A key-set test that
	// derives its expectation from the code proves nothing; 029's banner_test.go asserted a defect
	// for exactly this reason.
	want := []string{
		"badges", "brand", "categoryKey", "compareAtAmount", "currency", "id", "imageUrl",
		"name", "priceAmount", "priceDropped", "savedAt", "savedPriceAmount", "verdict",
	}
	sort.Strings(want)
	require.Equal(t, want, got)
}

// TestSavedItemDoesNotCarryAvailable pins the deliberate OMISSION.
//
// ⚠ `available` is the field this whole feature exists to replace: a boolean derived from catalogue
// status alone, which reported a product buyable while checkout refused it at the shopper's address.
// `verdict` supersedes it. If `available` ever reappears here, two fields answer the same question and
// a client will eventually render the wrong one — so its ABSENCE is a requirement, not an oversight.
func TestSavedItemDoesNotCarryAvailable(t *testing.T) {
	var m map[string]json.RawMessage
	require.NoError(t, json.Unmarshal([]byte(SAVED_ITEM_WIRE_JSON), &m))

	require.NotContains(t, m, "available",
		"a saved item reports `verdict`, never the catalogue-status boolean that lied")
}

// TestPriceDroppedIsOmittedWhenFalse pins FR-044's asymmetry on the wire.
//
// ⚠ There is deliberately no `priceRose`. The current price is always present, so nothing is
// concealed — but a rise is not actionable, and badging it would add noise to the one signal a
// watchlist exists to carry. `omitempty` is the mechanism, and it is load-bearing.
func TestPriceDroppedIsOmittedWhenFalse(t *testing.T) {
	d := wireFixture()
	d.PriceDropped = false

	raw, err := json.Marshal(d)
	require.NoError(t, err)

	var m map[string]json.RawMessage
	require.NoError(t, json.Unmarshal(raw, &m))
	require.NotContains(t, m, "priceDropped", "absent means no drop")
	require.NotContains(t, m, "priceRose", "this field must never exist")
}

// TestMembershipCountIsAWholeNumberOnTheWire is the 027 R13 guard.
//
// ⚠ `count` is a WireInt in the TS contract precisely so the generated Kotlin is a Long and not a
// Double. If Kotlin ever emits `1.0` here, Go's encoding/json REFUSES it into an int and the request
// 422s — invisibly, because unit tests on either side would still pass.
// ⚠ JSONEq, not string equality, and deliberately so. Go's encoding/json emits keys in STRUCT FIELD
// order (`productIds`, then `count`); quicktype sorts the generated Kotlin data class alphabetically
// (`count`, then `productIds`). Both encodings are valid and byte-different. Key order carries no
// meaning in JSON, so pinning it across two serialisers would pin an accident of each and break the
// day either tool changed. The literal is the shared DECODE fixture; what is asserted about EMISSION
// is the thing that actually broke in 027 — the numeric literal.
func TestMembershipCountIsAWholeNumberOnTheWire(t *testing.T) {
	raw, err := json.Marshal(membershipDTO{ProductIDs: []string{"9f2c1d4e-0000-0000-0000-000000000001"}, Count: 1})
	require.NoError(t, err)
	require.JSONEq(t, SAVED_MEMBERSHIP_WIRE_JSON, string(raw))
	require.Contains(t, string(raw), `"count":1`)
	require.NotContains(t, string(raw), `"count":1.0`)

	// And the reverse direction: Go must accept what Kotlin sends.
	var back membershipDTO
	require.NoError(t, json.Unmarshal([]byte(SAVED_MEMBERSHIP_WIRE_JSON), &back))
	require.Equal(t, 1, back.Count)
}

// TestGoRejectsAFloatCount is the negative proof that the guard above can fail.
func TestGoRejectsAFloatCount(t *testing.T) {
	var back membershipDTO
	err := json.Unmarshal([]byte(`{"productIds":[],"count":1.0}`), &back)
	require.Error(t, err,
		"this is the 027 defect exactly — if this ever stops erroring, the WireInt annotation has "+
			"been lost and Kotlin can silently start sending floats")
}

// TestEmptyListDecodesAsEmptyNotAFailure — a shopper with nothing saved is a normal state.
func TestEmptyListDecodesAsEmptyNotAFailure(t *testing.T) {
	var items []savedItemDTO
	require.NoError(t, json.Unmarshal([]byte(`[]`), &items))
	require.Empty(t, items)
}

// TestVerdictVocabularyIsClosed pins the five values both languages must agree on.
//
// ⚠ These are not free-form strings. Each implies a different next action for the shopper, and a
// sixth value appearing on the wire would reach a client that cannot render it.
func TestVerdictVocabularyIsClosed(t *testing.T) {
	require.Equal(t, []string{
		"purchasable",
		"temporarily_unavailable",
		"not_delivered_to_your_area",
		"no_longer_sold",
		"not_yet_determined",
	}, []string{
		VerdictPurchasable, VerdictTemporarilyOut, VerdictNotDeliveredHere,
		VerdictNoLongerSold, VerdictNotYetDetermined,
	})
}

// TestOrderLineCarriesProductID is the Go half of FR-008's dependency.
//
// ⚠ The order line has carried `productId` on the wire since 019; the MOBILE DOMAIN MODEL dropped it,
// so a shopper could read a past order and save nothing from it. That is the same
// mapper-discards-what-the-backend-sends shape that hid `brand` and `badges` on the saved list — and
// research R12 wrongly concluded the CONTRACT was the blocker. It was not; the field needed mapping.
// This pins the wire half so a future change cannot quietly remove what the mobile side now relies on.
func TestOrderLineCarriesProductID(t *testing.T) {
	const orderItemWire = `{"productId":"9f2c1d4e-0000-0000-0000-000000000001","productName":"Free Range Eggs 12pk","unitPriceAmount":"6.50","quantity":2,"lineSubtotalAmount":"13.00"}`

	var m map[string]json.RawMessage
	require.NoError(t, json.Unmarshal([]byte(orderItemWire), &m))
	require.Contains(t, m, "productId",
		"a product you cannot identify is a product you cannot save (FR-008)")
}

File: apis/edge-api/customer/src/customer/wire-contract.test.ts
import { readFileSync } from "node:fs"
import { join } from "node:path"
import { describe, expect, it } from "vitest"

import { toDTO } from "./model"
import type { CustomerRow } from "./model"

/**
 * THE CROSS-LANGUAGE WIRE CONTRACT for `CustomerDTO` (034 T033).
 *
 * ⚠ WHY A BYTE COMPARISON AND NOT A TYPE CHECK.
 *
 * TypeScript and Kotlin agree at compile time only about types they each derive from the same
 * source — and `Dto.kt` is GENERATED, so a rename in `customer.ts` regenerates the Kotlin happily and
 * both sides still compile. What breaks is the SERIALISED NAME on the wire, silently, at runtime, in
 * production only. Feature 027 lost days to exactly this class: Kotlin serialised a quantity as
 * `1.0`, Go refused it into an `int`, and every unit test on both sides passed because the fakes
 * spoke the same language at both ends and never crossed the wire.
 *
 * Feature 028 built the first such test (`wire_contract_test.go` + `BannerWireContractTest.kt`) and
 * proved it by breaking it two ways. This is the same shape for the customer record, which 034 has
 * just widened with two new fields.
 */

const row: CustomerRow = {
  id: "c-1",
  cognito_sub: "sub-1",
  email: "shopper@example.com",
  given_name: "Janith",
  family_name: "Madarasinghe",
  phone: "0400 000 000",
  status: "active",
  closure_state: "open",
  has_password: true,
  email_delivery: null,
  password_updated_at: new Date("2026-08-01T00:00:00Z"),
  created_at: new Date("2026-07-14T00:00:00Z"),
  updated_at: new Date("2026-07-14T00:00:00Z"),
}

describe("CustomerDTO wire contract", () => {
  /**
   * ⚠ THE KEY SET IS PINNED EXPLICITLY, and written from the CONTRACT rather than from the struct.
   *
   * Feature 033 recorded the failure mode this avoids: a key-set test passed because its expectation
   * had been written by reading the implementation, so it asserted only that the code agreed with
   * itself. This list is the wire shape the clients are entitled to.
   */
  it("emits exactly the agreed keys, in the agreed spelling", () => {
    expect(Object.keys(toDTO(row)).sort()).toEqual(
      [
        "closureState",
        "createdAt",
        "email",
        "emailDelivery",
        "familyName",
        "givenName",
        "hasPassword",
        "id",
        "passwordUpdatedAt",
        "phone",
        "status",
      ].sort(),
    )
  })

  /** ⚠ `cognito_sub` is an internal join key and must never reach the wire. */
  it("never leaks the identity-provider subject", () => {
    const json = JSON.stringify(toDTO(row))
    expect(json).not.toContain("cognito")
    expect(json).not.toContain("sub-1")
  })

  /**
   * ⚠ The generated Kotlin must declare the same property names. A silent rename in `customer.ts`
   * regenerates `Dto.kt` and both sides compile — this is what notices.
   */
  it("matches the generated Kotlin contract, field for field", () => {
    const dtoKt = readFileSync(
      join(__dirname, "../../../../../packages/shared-types/contract/Dto.kt"),
      "utf8",
    )

    // The `data class CustomerDTO ( … )` block.
    const block = /data class CustomerDTO \(([\s\S]*?)\n\)/.exec(dtoKt)
    expect(block, "CustomerDTO not found in the generated Kotlin — did the aggregator drop it?")
      .not.toBeNull()

    const body = block![1]!
    for (const key of Object.keys(toDTO(row))) {
      expect(body, `Kotlin CustomerDTO is missing "${key}"`).toContain(`val ${key}`)
    }
  })

  /** Dates cross the wire as ISO-8601 strings, never as epoch numbers or Date objects. */
  it("serialises timestamps as ISO-8601 strings", () => {
    const dto = toDTO(row)
    expect(dto.createdAt).toBe("2026-07-14T00:00:00.000Z")
    expect(dto.passwordUpdatedAt).toBe("2026-08-01T00:00:00.000Z")
  })

  /** `null` is a legitimate, complete value for each of these — not a gap to be defaulted away. */
  it("keeps null as null rather than inventing a value", () => {
    const dto = toDTO({
      ...row,
      given_name: null,
      family_name: null,
      phone: null,
      password_updated_at: null,
    })
    expect(dto.givenName).toBeNull()
    expect(dto.familyName).toBeNull()
    expect(dto.phone).toBeNull()
    expect(dto.passwordUpdatedAt).toBeNull()
  })
})
